chore: lint using putout

coderaiser opened the pull request:
After successfully merged linting sessions [part1](https://github.com/xtermjs/xterm.js/pull/2953), [part2](https://github.com/xtermjs/xterm.js/pull/3133) and [part3](https://github.com/xtermjs/xterm.js/pull/3269), the time is come for part 4 :).

As usual, any rule can be disabled.
Command used:

```sh
putout {src,addons}/**/*.ts --fix
```

Applied rules:
- [apply-as-type-assertions](https://github.com/coderaiser/putout/tree/v17.4.0/packages/plugin-apply-as-type-assertions)
- [simplify-ternary](https://github.com/coderaiser/putout/tree/v17.4.0/packages/plugin-simplify-ternary)
- [remove-useless-async](https://github.com/coderaiser/putout/tree/v17.4.0/packages/plugin-remove-useless-async)
- [remove-useless-operand](https://github.com/coderaiser/putout/tree/v17.4.0/packages/plugin-remove-useless-operand)

Current config for [putout v17](https://github.com/coderaiser/putout):

```json
{
    "rules": {
        "apply-shorthand-properties": "on",
        "apply-destructuring": "off",
        "apply-numeric-separators": "off",
        "convert-assignment-to-comparison": "off",
        "convert-apply-to-spread": "off",
        "convert-math-pow": "off",
        "convert-for-to-for-of": "off",
        "convert-template-to-string": "off",
        "strict-mode": "off",
        "remove-useless-spread/object": "off",
        "remove-useless-array-constructor": "off",
        "remove-boolean-from-assertions": "off",
        "remove-iife": "off",
        "remove-console": "off",
        "remove-unused-variables": "off",
        "remove-useless-variables": "off",
        "merge-if-statements": "off",
        "promises/add-missing-await": "off"
    },
    "match": {
        "SerializeAddon.benchmark.ts": {
            "remove-unused-expressions": "off"
        }
    },
    "plugins": [
        "apply-shorthand-properties"
    ]
}
```

Tyriar commented on src/common/Clone.ts:
```diff
@@ -16,7 +16,7 @@ export function clone<T>(val: T, depth: number = 5): T {
   for (const key in val) {
     // Recursively clone eack item unless we're at the maximum depth
-    clonedObject[key] = depth <= 1 ? val[key] : (val[key] ? clone(val[key], depth - 1) : val[key]);
+    clonedObject[key] = val[key] || clone(val[key], depth - 1);
```
Looks like this change is a little too aggressive and changes the behavior

coderaiser replied:
About this code, we can rewrite it like this:

```js
const clone = () => {
    if (depth <= 1) {
        return val[key];
    } else {
        if (val[key]) {
            clone(val[key], depth - 1);
       } else {
           return val[key];
      }
    }
}
```

And simplify like this:

```js
const clone = (val, key, depth) => {
    if (val[key] && depth > 1)
       return clone(val[key], depth - 1);
    
    return val[key];
}
```

Yes, you absolutely right!

Just fixed `simplify-ternary`, will update PR.

Tyriar commented on src/browser/Terminal.ts:
```diff
@@ -1014,7 +1014,7 @@ export class Terminal extends CoreTerminal implements ITerminal {
     this._keyDownHandled = false;
 
-    if (this._customKeyEventHandler && this._customKeyEventHandler(event) === false) {
+    if (this._customKeyEventHandler && !this._customKeyEventHandler(event)) {
```
Not sure about this rule, typically I would explicitly use `false` when I think it's called for. This particular function is provided by embedders so it's a little more explicit when checking - perhaps internally it should return `unknown` 🤔

coderaiser replied:
OK, I'll better disable the rule `remove-boolean-from-assertions`

Tyriar commented on addons/xterm-addon-serialize/src/SerializeAddon.ts:
```diff
@@ -105,7 +105,7 @@ class StringSerializeHandler extends BaseSerializeHandler {
 
   protected _beforeSerialize(rows: number, start: number, end: number): void {
-    this._allRows = new Array<string>(rows);
+    this._allRows = Array<string>(rows);
```
Not a fan of removing the `new`

coderaiser replied:
Not problem, at all, I'll disable `remove-useless-array-constructor` :).

coderaiser commented:
Just disabled rules:
-  [remove-useless-array-constructor](https://github.com/coderaiser/putout/tree/v17.4.0/packages/plugin-remove-useless-array-constructor)
- [remove-boolean-from-assertions](https://github.com/coderaiser/putout/tree/v17.4.0/packages/plugin-remove-boolean-from-assertions)

And fixed:
- [simplify-ternary](https://github.com/coderaiser/putout/tree/v17.4.0/packages/plugin-simplify-ternary)(`@putout/plugin-simplify-ternary@v2.5.1`)

And updated `PR`, thanks a lot for quick response @Tyriar :)!

Tyriar commented on src/common/Clone.ts:
```diff
@@ -16,7 +16,7 @@ export function clone<T>(val: T, depth: number = 5): T {
   for (const key in val) {
     // Recursively clone eack item unless we're at the maximum depth
-    clonedObject[key] = depth <= 1 ? val[key] : (val[key] ? clone(val[key], depth - 1) : val[key]);
+    clonedObject[key] = depth <= 1 ? val[key] : (val[key] || clone(val[key], depth - 1));
```
This still isn't quite right as it clones the key if it's there

coderaiser replied:
You right! Fixed [simplify-ternary](
https://github.com/coderaiser/putout/tree/master/packages/plugin-simplify-ternary )([@putout/plugin-simplify-ternary@2.5.2](https://github.com/coderaiser/putout/commit/1ab404f33bf8ae89ab2933868f810326292aa622#diff-0ce4998d3f3a3d0ab4c56865edd5c0733ff26196f8c1f39bf14ca27e5e1386bd)).

Tyriar commented on src/browser/TestUtils.test.ts:
```diff
@@ -415,13 +415,13 @@ export class MockRenderService implements IRenderService {
     return true;
   }
-  public getJoinedCharacters(row: number): [number, number][] {
+  public getJoinedCharacters(row: number): Array<[number, number]> {
```
We prefer [] to Array<>, I think this is an eslint rule but not sure if we use it

coderaiser replied:
Looks like it's related to https://github.com/xtermjs/xterm.js/commit/8c65fcacb37e729da5f4c0307065bbe3b89252f2, maybe I had old `master` 🤔

jerch replied:
That was a weirdo before - we an array rule in place, but the older eslint version did not care much about it. After upgrading I got several linting warnings, and had to change the rule in 8c65fca to pass without warnings.

jerch commented on addons/xterm-addon-serialize/test/SerializeAddon.api.ts:
```diff
@@ -487,9 +487,9 @@ function newArray<T>(initial: T | ((index: number) => T), count: number): T[] {
   for (let i = 0; i < array.length; i++) {
     if (typeof initial === 'function') {
-      array[i] = (<(index: number) => T>initial)(i);
+      array[i] = (initial as (index: number) => T)(i);
```
@Tyriar Do we have a preference here? Could/should this be an eslint rule?

I admit, that I like the `xy as Z` more than `<Z>xy`, its imho easier to grasp with less syntax overhead (makes me less thinking). But thats just me.

coderaiser replied:
The thing is `Babel` can’t parse code that have both `react` and `typescript` enabled, for `xterm.js` it’s not a thing, but [new construction is preferable](https://basarat.gitbook.io/typescript/type-system/type-assertion#as-foo-vs-less-than-foo-greater-than).

Tyriar replied:
I prefer `as` more, more readable and a little easier to type. Looks like there is an eslint rule https://github.com/xtermjs/xterm.js/issues/3359

coderaiser replied:
Yes, this is the same rule, with the only difference: [it cannot fix](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/consistent-type-assertions.ts) :). Anyways  it will keep codebase in a clean state warning developers about assertions type used, so I think it is useful.

Tyriar replied:
Yeah I noticed that when I did the PR ☹️

jerch commented on addons/xterm-addon-webgl/src/GlyphRenderer.ts:
```diff
@@ -263,7 +263,7 @@ export class GlyphRenderer {
         // from bg. This is needed since the inverse fg color should be based on the original bg
         // color, not on the selection color
-        fg = (fg & ~(Attributes.CM_MASK | Attributes.RGB_MASK | FgFlags.INVERSE));
+        fg &= ~(Attributes.CM_MASK | Attributes.RGB_MASK | FgFlags.INVERSE);
```
:+1: for shorthand notation where possible.

Minor sidenote - in some older engines those were actually slower than the explicit long versions.

jerch commented on src/browser/Linkifier.test.ts:
```diff
@@ -210,7 +210,7 @@ describe('Linkifier', () => {
         linkifier.registerLinkMatcher(/test/, () => assert.fail(), {
           validationCallback: (url, cb) => {
-            count += 1;
+            ++count;
```
@coderaiser Is there a reason for `++count` instead of `count++`? In C I kinda always use the latter, if I dont bind the result, which leads to same ASM code, but always use the first in loops (which used to make a difference in old compilers, not anymore though). Is there a difference in JS?

Tyriar replied:
I always prefer the postfix as well, with the only exception being if I actually need the updated value

coderaiser replied:
There is only one reason: you see what’s going on the variable in the first place, so you can understand it faster:

`aVeryLongVariable++`

Anyways, it’s not a problem, I’ll add an option or change to postfix and disable a rule.

coderaiser replied:
Just updated

jerch commented on src/common/InputHandler.test.ts:
```diff
@@ -77,7 +77,7 @@ describe('InputHandler', () => {
       bufferService.reset();
     });
-    it('SL (scrollLeft)', async () => {
+    it('SL (scrollLeft)', () => {
```
Eww, guess I created a code smell here:

Your tool correctly found the asyncs to be stripped. The code smell happens at those `parseP` calls, those should have `await` in front to be consistent on code level. Created #3347 for that. (You can keep the rule as it is, gonna re-add the async keywords over there.)

jerch reviewed (commented):
Just some comments/Questions from my side.

Tyriar commented on src/browser/Terminal2.test.ts:
```diff
@@ -105,7 +105,7 @@ function formatError(input: string, output: string, expected: string): string {
     let counter = start || 0;
     return (s: string): string => {
-      counter += 1;
+      ++counter;
```
postfix?

coderaiser replied:
done ✅

Tyriar reviewed (approved):
Looks good, let's wait until https://github.com/xtermjs/xterm.js/pull/3360 is merged to merge this in since it fixes more `as` problems and also some typing bugs (like setTimeout instead of window.setTimeout)
